cloud/openssh/config_test.go

```diff
@@ -69,143 +69,227 @@ func TestHostOrMatchRegex(t *testing.T) {
-			// doesn't exist.
-			if errors.Is(err, fs.ErrNotExist) {
-				return nil
-			}
 			return err
 		}
 		if path == "." {
 			return nil
 		}
+		fsPathsEqual(t, got, want, path)
+		return nil
+	})
+	if err != nil {
+		t.Fatal("got error checking if file systems are equal:", err)
+	}
 
-		info, err := fs.Stat(srcFS, path)
+	err = fs.WalkDir(want, ".", func(path string, _ fs.DirEntry, err error) error {
 		if err != nil {
 			return err
 		}
-		if info.IsDir() {
-			return os.Mkdir(filepath.Join(dstDir, path), info.Mode().Perm())
-		}
-		data, err := fs.ReadFile(srcFS, path)
-		if err != nil {
-			return err
+		if path == "." {
+			return nil
 		}
-		return os.WriteFile(filepath.Join(dstDir, path), data, info.Mode().Perm())
+		fsPathsEqual(t, got, want, path)
+		return nil
 	})
 	if err != nil {
-		t.Fatal("got error duplicating testdata dir:", err)
+		t.Fatal("got error checking if file systems are equal:", err)
 	}
-	return dstDir
 }
 
-func dirsEqual(t *testing.T, gotDir, wantDir string) {
-	var perms map[string]fs.FileMode
-	if b, err := os.ReadFile(filepath.Join(wantDir, "perms.json")); err == nil {
-		v := make(map[string]string)
-		if err := json.Unmarshal(b, &v); err == nil {
-			perms = make(map[string]fs.FileMode, len(v))
-			for path, modeStr := range v {
-				mode, err := strconv.ParseUint(modeStr, 8, 32)
-				if err != nil {
-					t.Fatalf("path %q has invalid permissions %q", path, modeStr)
-				}
-				perms[path] = fs.FileMode(mode)
-			}
+func fsPathsEqual(t *testing.T, gotFS fs.FS, wantFS fs.FS, path string) {
+	t.Helper()
+
+	gotInfo, err := fs.Stat(gotFS, path)
+	if err != nil {
+		t.Errorf("got fs.Stat(gotFS, %q) error: %v", path, err)
+	}
+	wantInfo, err := fs.Stat(wantFS, path)
+	if err != nil {
+		t.Errorf("got fs.Stat(wantFS, %q) error: %v", path, err)
+	}
+	if gotInfo == nil || wantInfo == nil {
+		return
+	}
+	if got, want := gotInfo.Mode().Perm(), wantInfo.Mode().Perm(); got != want {
+		t.Errorf("got %q permissions %s, want %s", path, got, want)
+	}
+	if gotInfo.IsDir() != wantInfo.IsDir() {
+		gotType, wantType := "file", "file"
+		if gotInfo.IsDir() {
+			gotType = "directory"
 		}
+		if wantInfo.IsDir() {
+			wantType = "directory"
+		}
+		t.Errorf("got a %s at path %q, want a %s", gotType, path, wantType)
 	}
-	wantFS := os.DirFS(wantDir)
-	gotFS := os.DirFS(gotDir)
-	err := fs.WalkDir(wantFS, ".", func(path string, entry fs.DirEntry, err error) error {
-		if err != nil {
-			// If the wantDir is missing then there's nothing to
-			// check.
-			if errors.Is(err, fs.ErrNotExist) {
+	if gotInfo.IsDir() || wantInfo.IsDir() {
+		return
+	}
+
+	gotBytes, err := fs.ReadFile(gotFS, path)
+	if err != nil {
+		t.Errorf("got fs.ReadFile(gotFS, %q) error: %v", path, err)
+	}
+	wantBytes, err := fs.ReadFile(wantFS, path)
+	if err != nil {
+		t.Errorf("got fs.ReadFile(wantFS, %q) error: %v", path, err)
+	}
+	diff := cmp.Diff(os.ExpandEnv(string(wantBytes)), os.ExpandEnv(string(gotBytes)))
```

Comment:
if the expandEnv is just for `$HOME`, I'd vote for leaving it out. What benefit do we get from it?

Reply:
I should probably comment this part.

When the test runs `SetupDevbox`, it uses a temp directory as a fake `$HOME`. This results in `~/.ssh/config` having absolute paths like:

    Include "/var/folders/79/1yc1ywp10w9f2xnr_rpp_ff00000gn/T/tmp.PnwhPb9h/.config/devbox/ssh"

Expanding environment variables lets us put `$HOME` in the golden files so that they can be compared with test output.